main.go

```diff
@@ -0,0 +1,517 @@
+	now := time.Now()
+	lf, err := setupLogFile()
+	if err != nil {
+		fmt.Printf("\nCan't set up log file: %v\n", err)
+		panic(fmt.Errorf("can't set up log file"))
+	}
+	defer close(lf)
+	project, err := getProject()
+	if err != nil {
+		fmt.Printf("\nCan't get project: %v\n", err)
+		panic(fmt.Errorf("can't get project"))
+	}
+	instance, err := getInstance(project)
+	if err != nil {
+		fmt.Printf("\nCan't get instance: %v\n", err)
+		panic(fmt.Errorf("can't get instance"))
+	}
+	fmt.Printf("WARNING: Please check that permissions for this instance are appropriate.\n")
+	fmt.Printf("Spanner manages access control at the database level, and the database\n")
+	fmt.Printf("created by HarbourBridge will inherit default permissions from this instance.\n")
+	fmt.Printf("All data written to Spanner will be visible to anyone who can access the\n")
+	fmt.Printf("created database. Note that PostgreSQL table-level and row-level ACLs\n")
+	fmt.Printf("are dropped during conversion since they are not supported by Spanner.\n")
+	f, n, err := getSeekable(os.Stdin)
+	if err != nil {
+		fmt.Printf("\nCan't get seekable input file: %v\n", err)
+		panic(fmt.Errorf("can't get seekable input file"))
+	}
+	defer f.Close()
+	conv := postgres.MakeConv()
+	dbName, err := getDatabaseName(now)
+	if err != nil {
+		fmt.Printf("\nCan't get database name: %v\n", err)
+		panic(fmt.Errorf("can't get database name"))
+	}
+	setFilePrefix(dbName)
+	firstPass(f, n, conv)
+	writeSchemaFile(conv, now, filePrefix+schemaFile)
+	db, err := createDatabase(project, instance, dbName, now, conv)
+	if err != nil {
+		fmt.Printf("\nCan't create database: %v\n", err)
+		panic(fmt.Errorf("can't create database"))
+	}
+	client, err := getClient(db)
+	if err != nil {
+		fmt.Printf("\nCan't create client for db %s: %v\n", db, err)
+		panic(fmt.Errorf("can't create Spanner client"))
+	}
+	_, err = f.Seek(0, 0)
+	if err != nil {
+		fmt.Printf("\nCan't seek to start of file (preparation for second pass): %v\n", err)
+		panic(fmt.Errorf("can't seek to start of file"))
+	}
+	rows := conv.Rows()
+	bw := secondPass(f, client, conv, rows)
+	report(bw, n, now, db, conv)
+}
+
+func report(bw *spanner.BatchWriter, bytesRead int64, now time.Time, db string, conv *postgres.Conv) {
+	fileName := filePrefix + reportFile
+	f, err := os.Create(fileName)
+	if err != nil {
+		fmt.Printf("Can't write out report file %s: %v\n", fileName, err)
+		fmt.Print("Writing report to stdout\n")
+		f = os.Stdout
+	} else {
+		defer f.Close()
+	}
+	w := bufio.NewWriter(f)
+	banner := fmt.Sprintf("Generated %s for db %s\n\n", now.Format("2006-01-02 15:04:05"), db)
+	w.WriteString(banner)
+	badWrites := bw.DroppedRowsByTable()
+	summary := postgres.GenerateReport(conv, w, badWrites)
+	w.Flush()
+	fmt.Printf("Processed %d bytes of pg_dump data (%d statements, %d rows of data, %d errors, %d unexpected conditions).\n",
+		bytesRead, conv.Statements(), conv.Rows(), conv.StatementErrors(), conv.Unexpecteds())
+	if f != os.Stdout {
```

Comment:
Actually summary is short.
And we do want to print it to stdout.

In the typical case, f is the report file, and we want to write summary to both the report file and to stdout.

However, in the special case where we couldn't open f, we write the report to stdout -- see the logic a few lines earlier. (Aside: its debatable whether this extra complexity is needed: we could just fail if we can't write the report, or just not generate the report. The motivation for the current logic is to try to continue if we can since this is intended to be a turnkey tool and the report file is pretty important.)

The reason for the check is that we want to avoid writing summary twice.

Reply:
Added a comment to clarify.